vision.py

```diff
@@ -48,45 +48,90 @@ def process_image(self, frame: np.ndarray) -> Optional[Tuple[float, float]]:
+        return self.get_values(contours, best_group, (frame.shape[1], frame.shape[0]))
 
-    def preprocess(self, frame):
-        """Do any preprocessing we might want to do on the frame
-        e.g. convert to hsv, threshold, erode/dilate
+    def preprocess(self, frame: np.ndarray) -> np.ndarray:
+        """Creates a mask of expected target green color
         """
-        return frame
-
-    def find_contours(self, frame):
-        """Finds the contours, see https://docs.opencv.org/4.x/d4/d73/tutorial_py_contours_begin.html
+        hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
+        mask = cv2.inRange(hsv, TARGET_HSV_LOW, TARGET_HSV_HIGH)
+        mask = cv2.erode(mask, ERODE_DILATE_KERNEL)
+        mask = cv2.dilate(mask, ERODE_DILATE_KERNEL)
+        return mask
+
+    def find_contours(self, mask: np.ndarray) -> np.ndarray:
+        """Finds contours on a grayscale mask
         """
-        return []
+        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
+        return contours
 
-    def filter_contours(self, contours):
-        """Takes contours and returns on ones that are likely to be vision tape
-        Checks things like size, width to height ratio
+    def filter_contours(self, contours: np.ndarray) -> np.ndarray:
+        """Unimplemented, returns identity
         """
-        return []
+        return contours
 
-    def group_contours(self, contours):
-        """Create potential groups out of the contours
-        Group based on distance, sizes and relative positions
+    def group_contours(self, contours: np.ndarray) -> List[List[int]]:
+        """Returs a nested list of contour indices grouped by relative distance
         """
-        pass
-
-    def rank_groups(self, groups):
+        # Build an adjacency list based on distance between contours relative to their area
+        connections = [[] for _ in contours]
+        for i, a in enumerate(contours):
+            a_com = a.mean(axis=0)[0]
+            for j, b in enumerate(contours[i + 1:]):
+                max_metric = max(cv2.contourArea(a), cv2.contourArea(b)) * METRIC_SCALE_FACTOR
+                b_com = b.mean(axis=0)[0]
+                d = np.abs(a_com - b_com)
+                metric = d[0] * d[0] + d[1] * d[1]
+                if (metric < max_metric):
+                    connections[i].append(j + i + 1)
+                    connections[j + i + 1].append(i)
+        # Breadth first search from each contour that wasn't yet assigned to a group to find its group
+        assigned = [False for _ in contours]
+        group = set()
+        groups = list()
+        for i, c in enumerate(connections):
+            if assigned[i]: continue
+            group = {i}
+            assigned[i] = True
+            horizon = set(c)
+            while (True):
+                if len(horizon) == 0:
+                    # Group completed
+                    groups.append(list(group))
```

Comment:
curiosity question: do we need each `group` to be a list where this is used? can we just return the set instead?